### ISSUE TITLE
Release alpha version with node fixes

### ISSUE THREAD

**Issue by hoodmane** [pyodide/pyodide#2403]
It would be good to release an alpha version with #2394 to help the node people test whether #2393 and related issues are fixed. We also have problems with webpack which probably aren't fixed, maybe we should try to improve the webpack situation first too.

I'm not sure how to version this stuff. I wouldn't normally want to include #2387 and #2391 in a patch release, but we could backport everything else. Or we could throw them into `0.20.1` they aren't humongous changes. Or we could plan to do a very small `v0.21.0` release.

 @alexmojaki @mnell @rth @ryanking13

**Comment by ryanking13** [comment]
I think `v0.20.1` would be good because I recently found that we didn't version `pyodide-build` `v0.20.0` correctly in PyPI (https://pypi.org/project/pyodide-build/#history), so we can bump the version together.

**Comment by hoodmane** [comment]
I think it would be good to release an alpha first e.g., `0.20.1a1` so that if there are more dumb mistakes with the node/webpage stuff we can iron those out first.

https://github.com/pyodide/pyodide/pull/2405 would also be good to include.

**Comment by ryanking13** [comment]
> I think it would be good to release an alpha first e.g., `0.20.1a1` so that if there are more dumb mistakes with the node/webpage stuff we can iron those out first.
> 
> #2405 would also be good to include.

I'm +1 with it

**Comment by alexmojaki** [comment]
Seems like this release could be made now? I'm keen to try it out.

**Comment by ryanking13** [comment]
I made an alpha release `v0.20.1a1`, though we didn't fix webpack situation yet... #2387 and #2391 are included as they have conflict with recent changes and I didn't want to mess up. Excluded commits are #2411, #2407, #2389, #2305.

@rth @hoodmane could you publish JS package to npm registry? I don't have access to our npm registry.

**Comment by rth** [comment]
Thanks! No worries, as long as it worked. I send you the invite.

**Comment by ryanking13** [comment]
I find out that I published the wrong stuff (TypeScript files, not JavaScript) in npm `pyodide@0.20.1-alpha.1`. 

It looks like npm does not allow removing and republishing the same version. So I deprecated `pyodide@0.20.1-alpha.1` for now. We should publish again in another version name (Maybe, `pyodide@0.20.1-alpha.2`?).

Sorry about this. FYI, this happened because I called `npm publish` in `src/js` not in `dist`, which was changed after #2394.

**Comment by rth** [comment]
No, worries, let's push a pyodide@0.20.1-alpha.2 and I think you should be able to yank the other one.

**Comment by ryanking13** [comment]
> No, worries, let's push a pyodide@0.20.1-alpha.2 and I think you should be able to yank the other one.

Pushed to `pyodide@0.20.1-alpha.2`.

It seems like unpublishing npm package is forbidden when there is a dependent package (https://github.com/npm/cli/issues/1686#issuecomment-840668668).

**Comment by alexmojaki** [comment]
> Sorry about the mistake.

Being able to make mistakes is the entire point of this issue! Thanks for making a release so that it can be tested. I tried it out and here's what I found:

- Many of the types/functions in `pyodide.d.ts` are declared but not exported, particularly `PyodideInterface` and its methods like `unpackArchive` and `setInterruptBuffer`.
- `unpackArchive` doesn't accept `ArrayBuffer` for the first argument, I think it should.
- I get a lot of errors about missing files in `pyodide/node_modules/ws/lib`, e.g:

```
ERROR in ../node_modules/pyodide/node_modules/ws/lib/receiver.js 3:21-38
Module not found: Error: Can't resolve 'stream' in '/home/alex/work/pyodide-worker-runner/node_modules/pyodide/node_modules/ws/lib'
Did you mean './stream'?
Requests that should resolve in the current directory need to start with './'.
Requests that start with a name are treated as module requests and resolve within module directories (node_modules).
If changing the source code is not an option there is also a resolve options called 'preferRelative' which tries to resolve these kind of requests in the current directory too.

BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

If you want to include a polyfill, you need to:
        - add a fallback 'resolve.fallback: { "stream": require.resolve("stream-browserify") }'
        - install 'stream-browserify'
If you don't want to include a polyfill, you can use an empty module like this:
        resolve.fallback: { "stream": false }
 @ ../node_modules/pyodide/node_modules/ws/wrapper.mjs 2:0-41 7:0-79
 @ ../node_modules/pyodide/pyodide.mjs 318:20-32
 @ ../lib/index.ts 13:0-38 17:11-22
 @ ./index.ts 11:0-39 34:27-40
```

**Comment by rth** [comment]
Thanks a lot for testing it @alexmojaki . Maybe @mneil would be interested in having a look.

**Comment by hoodmane** [comment]
I think the webpack issue is hard. Ideally webpack should not try to bundle `pyodide.asm.js`. But `pyodide.js` is the output of rollup so it's not clear how to get a `/* webpackIgnore: true */` comment into it (ideally at the same time we could remove the Microsoft copyright statement...)

**Comment by hoodmane** [comment]
These are separate files because Emscripten generates them that way. But we could switch to rolling up (or hopefully esbuilding soon) `pyodide.asm.js` into `pyodide.js`. I think Emscripten's loading logic expects `pyodide.asm.data` to be a separate file with that name. Also, it is encoded in binary which is not possible in a JS file. So in any case that file should stay separate. I doubt any bundler is smart enough to understand how Emscripten tries to load it anyways.

We really want `pyodide.js` to be code split from the rest of a webpack bundle. Running webpack over the Emscripten JavaScript output is not a good idea IMO. Even without the resolution problems, it could still create confusing problems.

**Comment by mneil** [comment]
Not the best for for bundle size but the data file could be bases64 encoded and inlined into the asm.js. THe benefit of that is loading / setup on the browser becomes simpler. Downside is you get about a 30% bump in filesize on the encoded file.

> Running webpack over the Emscripten JavaScript output is not a good idea IMO

Are there specific concerns about this? One other issue I have right now with the pyodide.js file is that it forces the use of some globals. It is also checking for "IS_NODE" in a way that fails in webpack since process, fs, require, can all be mocked or replaced entirely causing pyodide.js to believe it is in node when it is in the browser.

Bringing pyodide.asm.js and pyodide.js into a single file (and potentially encoding the data file as well into a single file) and removing the hard requirement of globals for the file system in particular would allow users like myself to choose their own filesystem replacement, their own fetch implementation, etc...

**Comment by hoodmane** [comment]
> bases64 encoded ... a 30% bump in filesize

Also an increase in load time and ram usage. See the [Emscripten docs](https://emscripten.org/docs/tools_reference/emcc.html#emcc-embed-file):

> Embedding files is much less efficient than [preloading](https://emscripten.org/docs/tools_reference/emcc.html#emcc-preload-file) them. You should only use it for small files, in small numbers.

I don't think we are interested in going this route.

> Are there specific concerns about this?

Emscripten tries to import ~4 different node modules only in node, not in browser. So these would need to be added to webpack externals, or we would need to patch Emscripten not to do this or... something? But also, it's 2.7 Mb of Javascript which has already been bundled and treeshaken and completely prepared for use by Emscripten, having webpack try to parse and optimize that again seems like it would be unnecessarily bad for build times. And it is completely possible webpack could mess things up.

Another thing to consider is that we get better load times by splitting these files the way we do, because `pyodide.js` is small and so it can initiate other requests we need for `packages.json`, `pyodide.asm.js`, and `pyodide_py.tar` all at the same time.

Also, splitting the files significantly improves the speed of partial rebuilds, particularly when only working on the Python source files or the Typescript source files and not the C code. rollup is really slow, though with esbuild it would be less of a problem.

You may be able convince us, I am not very knowledgeable about the Javascript ecosystem.

>  would allow users like myself to choose their own filesystem replacement, their own fetch implementation, etc...

We are interested in allowing those features, but I'm not sure the normal import system is the best way. We could for instance add a loading setting that tells Pyodide whether it is in the browser or not and overrides the guess.

**Comment by bollwyvl** [comment]
> add a loading setting that tells Pyodide whether it is in the browser

:+1: explicit is better than implicit, yadda yadda: if `loadPyodide`'s `ConfigType` included a `fetchImpl`, etc. it would be _very_ interesting for the jupyterlite case, where we're already implementing a fake web server that has interesting things that should appear to be coming from it, and various other servers. Among them: move some of the `micropip` customization stuff out of the python-in-js side, and work with it a little more directly in the bootstrap lifecycle before we get access to the pyodide instance.

**Comment by bollwyvl** [comment]
> 30% bump in filesize on the encoded file.

oof! no, thanks!

**Comment by alexmojaki** [comment]
> I think the webpack issue is tricky. Ideally webpack should not try to bundle `pyodide.asm.js`. But `pyodide.js` is the output of rollup so it's not clear how to get a `/* webpackIgnore: true */` comment into it (ideally at the same time we could remove the Microsoft copyright statement...)

@hoodmane are you saying that errors like `Module not found: Error: Can't resolve 'stream' in '/home/alex/work/pyodide-worker-runner/node_modules/pyodide/node_modules/ws/lib'` come from trying to bundle `pyodide.asm.js`, which should only be loaded at runtime?

**Comment by hoodmane** [comment]
Yeah more or less. The other issue is that webpack thinks that node apis should be shimmed in the browser but most modules we load are node shims for browser apis.

**Comment by hoodmane** [comment]
Well I think the issue is that we didn't deploy all the packages to npm to reduce download size -- we didn't want to deploy 200+ mb to npm. Try using:
```js
pyodide.loadPackage({indexURL : "https://cdn.jsdelivr.net/pyodide/v0.20.1a1/full/" });
```

**Comment by rth** [comment]
> I think the issue is that we didn't deploy all the packages to npm to reduce download size

I think that's what we want, though? So there is an issue in v0.20.1a1 that packages.json doesn't include all packages (while it should), however aside from that last time I tried to use it in node it does download and cache the packages `node_modules` when needed (even, without specifying the `indexURL`). That was without webpack though. So if packages need to be present in node_modules already, maybe we could have some CLI tool (or webpack plugin to do that) but IMO it would be better not to bundle all packages into the pyodide NPM package as most users shouldn't need that, and this apporach doesn't scale as we get more packages.

But then again I have little experience with the JS packaging ecosystem, so if someone is sure there is no way around it, we can certainly change it.

**Comment by ryanking13** [comment]
Thanks for the experiments and for opening a discussion @mneil. Maybe we need to use a separate issue for this conversation.

> Document the need to download these individually

I was thinking of providing a simple command (#2402) to install wheel files from CDN, which can be done automatically or manually. I'm not sure how these downloaded packages can be easily shipped in bundlers though.

**Comment by wenq1** [comment]
Two issues still with 1.20.1-a2:

1. Doc says it should be imported as below
```
let pyodide_pkg = await import("pyodide/pyodide.js");
```

However, in reality if using ESM, one should import as below:

```
import    { loadPyodide }                                       from 'pyodide/pyodide.mjs';
```

Note that the filename should be 'pyodide.mjs' not 'pyodide.js', and there's no default export from 'pyodide.mjs'

2. In nodejs 16, it fails with below

```
RuntimeError: abort(undefined). Build with -s ASSERTIONS=1 for more info.
    at process.abort (/Volumes/SecureVol1/Project/NewToken/kean-server/node_modules/pyodide/pyodide.asm.js:10:234679)

```
Code used is:

```
        let     artifacts_dir           = path.join (project_root, 'node_modules/pyodide');
        let     pyodide         = await loadPyodide ({
                indexURL:       artifacts_dir,
        });

```

**Comment by hoodmane** [comment]
Thanks for the info about the docs being wrong. 

In node v16.14.0 I don't get `RuntimeError: abort(undefined)`:
```js
import { loadPyodide } from "pyodide/pyodide.mjs";

async function main(){
    let pyodide = await loadPyodide({"indexURL" : "./node_modules/pyodide"});
    pyodide.runPython("print('1+1 is', 1+1)");
    console.log(pyodide.version);
}
main();
```
prints:
```
warning: no blob constructor, cannot create blobs with mimetypes
warning: no BlobBuilder
Loading distutils
Loaded distutils
Python initialization complete
1+1 is 2
0.20.1a1
```

**Comment by wenq1** [comment]
For the same code, i get
```
warning: no blob constructor, cannot create blobs with mimetypes
warning: no BlobBuilder
Loading distutils
Loaded distutils
Python initialization complete
undefined
/Volumes/SecureVol1/Project/NewToken/kean-server/node_modules/pyodide/pyodide.asm.js:10
var Module=typeof _createPyodideModule!=="undefined"?_createPyodideModule:{};var readyPromiseResolve,readyPromiseReject;Module["ready"]=new Promise(function(resolve,reject){readyPromiseResolve=resolve;readyPromiseReject=reject});if(!Module.expectedDataFileDownloads){Module.expectedDataFileDownloads=0}Module.expectedDataFileDownloads++;(function(){var loadPackage=function(metadata){var PACKAGE_PATH="";if(typeof window==="object"){PACKAGE_PATH=window["encodeURIComponent"](window.location.pathname.toString().substring(0,window.lo

...

xit"](code);ABORT=true}quit_(code,new ExitStatus(code))}if(Module["preInit"]){if(typeof Module["preInit"]=="function")Module["preInit"]=[Module["preInit"]];while(Module["preInit"].length>0){Module["preInit"].pop()()}}var shouldRunNow=true;if(Module["noInitialRun"])shouldRunNow=false;run();
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            

RuntimeError: abort(undefined). Build with -s ASSERTIONS=1 for more info.
    at process.abort (/root/project_root/node_modules/pyodide/pyodide.asm.js:10:234679)
    at process.emit (node:events:526:28)
    at emit (node:internal/process/promises:140:20)
    at processPromiseRejections (node:internal/process/promises:274:27)
    at processTicksAndRejections (node:internal/process/task_queues:97:32)


```

**Comment by hoodmane** [comment]
> I'm using an Intel version of node on on M1 mac... maybe that's an issue

Hard to know. It is even possible that there could be a regression from v16.14.0 to v16.14.2 though this is probably even less likely. Do you have nvm? Do other versions of node have the problem too?

**Comment by hoodmane** [comment]
We appreciate the bug report even if it isn't obvious how to reproduce it.

**Comment by wenq1** [comment]
Thanks I can confirm that with node arm version, it works as expected. It might well just be the issue with apple silicion support of wasm. I'll do some more tests.

```
drwxr-xr-x  10 wc  staff   320 May 22 12:58 node_modules
-rw-r--r--   1 wc  staff  4893 May 22 12:58 package-lock.json
-rw-r--r--   1 wc  staff  1204 May 22 13:00 package.json
-rw-r--r--   1 wc  staff  1993 May 22 13:00 test-pyodide.js
wc@wcAir test-pyodide % node test-pyodide.js 
warning: no blob constructor, cannot create blobs with mimetypes
warning: no BlobBuilder
Loading distutils
Loaded distutils
Python initialization complete
1+1 is 2
0.20.1a1
wc@wcAir test-pyodide % which node
/opt/homebrew/opt/node@16/bin/node
wc@wcAir test-pyodide % file $(which node)
/opt/homebrew/opt/node@16/bin/node: Mach-O 64-bit executable arm64
```

**Comment by wenq1** [comment]
Does numpy with nodejs version?

I get the following:
```
RuntimeError: abort(Error: No known package with name 'numpy'). Build with -s ASSERTIONS=1 for more info.

```

**Comment by TheLostLambda** [comment]
Sorry @hoodmane for pulling this issue back up out of the grave, but your "./node_modules/pyodide" option for the indexURL is something I spent literally hours trying to find.

> ```js
> import { loadPyodide } from "pyodide/pyodide.mjs";
> 
> async function main(){
>     let pyodide = await loadPyodide({"indexURL" : "./node_modules/pyodide"});
>     pyodide.runPython("print('1+1 is', 1+1)");
>     console.log(pyodide.version);
> }
> main();
> ```

My only problem is that that breaks when the application is bundled for deployment (the `node_modules` directory isn't available) — do you have any idea how to actually ship Pyodide with a bundler? Or should I give up and just pull it in from a CDN?

**Comment by hoodmane** [comment]
Well there's the https://github.com/pyodide/pyodide-webpack-plugin which I believe can bundle Pyodide.
